### ISSUE TITLE
startMode="AlwaysRunning" for inProcess ANCM

### ISSUE THREAD

**Issue by shirhatti** [aspnet/AspNetCore#3849]
There was an earlier issue asking for this (https://github.com/aspnet/AspNetCoreModule/issues/29) which we declined to do when ANCM was only out of process.

At the minimum, we need to ensure ANCM doesn't break with AlwaysRunning for 2,1. I suspect it might **just work**, but @jkotalik and @pan-wang know better.

**Comment by Looooooka** [comment]
Moved to 2.2 but autostart doesn't work.
Since there is an IISIntegration http request that happens when something goes wrong, instructing the netcore app to shutdown(guessing there's a route registered to handle this event... because clearly a http://host/iisintegration request can be logged) why not just add similar request for when an application pool is recycled or an application is stopped/started/restarted and then just raise the events already in place for these "scenarios":
          applicationLifetime.ApplicationStarted
                applicationLifetime.ApplicationStopped;
            applicationLifetime.ApplicationStopping
This would honestly enable everyone to wire up their app from those points the way they needed to.

**Comment by dotnetup** [comment]
After setup `startMode=AlwaysRunning` on the AppPool and `preloadEnabled=true` on the site, ASP.NET Core app be able to start automatically without any request. 

https://blogs.msdn.microsoft.com/vijaysk/2012/10/11/iis-8-whats-new-website-settings/

**Comment by carlowahlstedt** [comment]
@jkotalik now that 2.2 is released, did the documentation get completed?

**Comment by Joelbear5** [comment]
How is this implemented when running Kestrel in Linux (without IIS as the reverse proxy server)?

**Comment by jkotalik** [comment]
@Looooooka with the release of 2.2, the module AspNetCoreModuleV2 should make startMode="AlwaysRunning" work. Please give it a try.

@carlowahlstedt we now have documentation on ANCM Inprocess in our docs. See https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/?view=aspnetcore-2.2#iis-options and https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/aspnet-core-module?view=aspnetcore-2.2

@Joelbear5 startMode="AlwaysRunning" is a feature of IIS and isn't implemented by Kestrel itself. If you are using a reverse proxy, there may be a feature to keep Kestrel alive.

**Comment by jkotalik** [comment]
@carlowahlstedt sorry for the incredibly late reply. <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel> is used to populate the web.config if there isn't one present in the project. If you have a web.config in your directory, it should always be used when you publish the application.

**Comment by YuliiaBabychKdd** [comment]
any updates? 
We have netcoreapp2.2, but still same issue:(

Any magic tricks or ideas how to fix? or when it will be fixed?

**Comment by BinaryPatrick** [comment]
We just wrote a tickle script
```
param($Uri)

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
Invoke-WebRequest -Uri $Uri | Select-Object -ExpandProperty StatusCode
```

**Comment by YuliiaBabychKdd** [comment]
@BinaryPatrick Thanks. Finished with same solution^^

**Comment by robbaman** [comment]
I've noticed that when you setup AlwaysRunning with a AspNet Core application (2.2) and I recycle the application (by clicking recycle in the IIS Manager) it correctly starts a new instance including the actual Kestrel-based application. This is also visible in the Windows Event Log with two events:

4-7-2019 12:04:24: Application 'D:\Sites\ShippingService\' started the coreclr in-process successfully.
4-7-2019 12:04:26: Application 'MACHINE/WEBROOT/APPHOST/SHIPPINGSERVICE' has shutdown.

Note how the Shutdown follows the start of a new instance. When the application pool lifecycle is managed/interrupted by an app_offline.html however, this does not work. Instead the logs are these:

4-7-2019 11:58:30: Application 'D:\Sites\ShippingService\' was recycled after detecting the app_offline file.
4-7-2019 11:58:33: Application 'MACHINE/WEBROOT/APPHOST/SHIPPINGSERVICE' has shutdown.

Notice a lack of the "started the coreclr in-process successfully" event in the logs. The Kestrel-application does indeed not start in this instance. This makes publishing a new version via Webdeploy.exe pretty hard. Once a single request comes in the site starts properly and everything works.

**Comment by unionthugface** [comment]
Sorry to reply to an old thread, but I am desperate. I have a .NET Core 2.2 app, in-process, hosted in IIS 10.  I have application pool set to AlwaysRunning with Timeout=0, Preload=true; however my app still stops every 29 hours when the application pool recycles, and then does not restart itself until someone pings the URL.  Back in the .NET Framework days I could write a bootstrapper class and hook into the Global.asax file/IIS ApplicationPreload settings; I haven't found any such thing for .NET Core and would so appreciate if someone could point me to one.  My office is going nuts because we moved a recurring jobs handler app to .NET Core, and now it won't work because of this application pool recycle thing; every 29 hours on the dot, it stops.  What am I missing?  If I go back to my boss and tell him we need to write a tickle script like someone suggested, he'll be upset.

**Comment by Looooooka** [comment]
First of all just set the managed pipeline to classic and net clr mode to no managed code. 
As you've noticed it's completely useless since the preload thing doesn't work as expected so loading the NET CLR is a giant waste of resources.

Instead change the Configure method in Startup.cs and register a method you wish to execute when the app is started using the ApplicationStarted functionality(you can handle the shutdown as well).
Then if you want to let's say "preload" a controller or anything so speed up requests just add some httpclient logic in that (OnStartUp) method and make a call to the website itself. Seems to work for us(since 2.0 and now on 3.1 as well).

`public void Configure(IHostApplicationLifetime applicationLifetime, IApplicationBuilder app)
        {
            applicationLifetime.ApplicationStarted.Register(OnStartUp);
}`

**Comment by Joelbear5** [comment]
My team has migrated away from dotnet core due to this issue. Even setting health check pings to an API only loads that controller. I think Azure might have a way to keep APIs "warm", but IMHO dotnet core completely failed us as a web server.

**Comment by Coruscate5** [comment]
For all those still experiencing this issue - after a lot of research, I determined that "new" servers did not have any issues autostarting .NET Core apps per the docs.

Based on ProcMon and Net call traces, it looks like this is a problem with IIS - it gets confused when multiple custom modules are installed after original install.

So, the solution was (somewhat) simple - Reinstall IIS completely (then install your custom .NET core runtime modules again after). Most of your settings will be saved - I wrote a logger in my test service to make sure the service kept running as well. 

Check my SO post: https://stackoverflow.com/questions/59905391/net-core-3-iis-application-initialization-doesnt-work

**Comment by guardrex** [comment]
Feedback on the IIS topic indicates that this might be the issue there. https://github.com/aspnet/AspNetCore.Docs/issues/16675

**Comment by ljani** [comment]
This seems a good issue to mention that if IIS has only Windows authentication enabled, then this won't work. You will need to enable anonymous authentication as well. I was able to figure this out using Failed Requests Tracing.

**Comment by Coruscate5** [comment]
@guardrex - I'm not sure if you have any responsibility over the IIS portion here, but, it appears that IIS eventually stops respecting the warm-up even after a reinstall and re-registering of modules.

Seems like some sort of internal IIS issue (or chaining issue), because it will load & run methods from warmup.dll, but then never actually start the dotnet process specifically for .NET Core. (Re)reinstalling IIS fixes it temporarily

**Comment by guardrex** [comment]
I just work on the docs @Coruscate5. Engineering (here) would need to follow up on your report.

**Comment by Coruscate5** [comment]
@guardrex - OK, thanks - given that this particular thread seems to have gone off the rails, I'll open a new issue under the aspnetcore repo when I have some time (given the strange nature of this happening "eventually" to IIS servers hosting .NET Core projects with background tasks, I'm not sure if I'll actually get any traction)

**Comment by unionthugface** [comment]
I posted this potential solution on @Coruscate5 's thread new issue #19509 above [https://github.com/dotnet/aspnetcore/issues/19509#issuecomment-683852466](https://github.com/dotnet/aspnetcore/issues/19509#issuecomment-683852466) but figured I'd post it here too:

> In my case, lack of Garbage Collecting was making the problem. I traced the memory usage and it got increased by any page load. I just added this to my .csproj file:
> `  <PropertyGroup>
>     <ServerGarbageCollection>true</ServerGarbageCollection>
>     <ConcurrentGarbageCollection>false</ConcurrentGarbageCollection>
>   </PropertyGroup>
> `
> and the problem got solved.

_Originally posted by @mobinseven in https://github.com/dotnet/aspnetcore/issues/11950#issuecomment-520295866_

I deployed this in my .NET Core 3.1 app on Friday, came back in this morning (Monday), and my app successfully shut itself down and restarted itself all weekend, no gaps in service.  

tl:dr; this appears to be a problem with .NET Core garbage collection.

Would love if anyone could chime in about why this works and/or any potential side effects.  I am too ignorant to make a guess, but my app is definitely working as expected with this code in place.

**Comment by MV10** [comment]
I'll be "that guy" and bump this ... is any of this targeted for any specific release?

Specifically, the `[IProcessHostPreloadClient](https://learn.microsoft.com/en-us/dotnet/api/system.web.hosting.iprocesshostpreloadclient?view=netframework-4.8.1)` option was only mentioned once way back in 2018, but it is still very different from the other options and very important because it happens without any restrictions or overhead associated with servicing a request.

We have some systems that have to cache data from very, very slow mainframe back-ends, and if they rely on requests for their initial load, those requests will time out. Those folks are now migrating their apps to ASP.NET Core (finally!) and all we have to guarantee pre-load are goofy scripts.

**Comment by MV10** [comment]
Sigh, another year goes by...

The pre-load scripts (ie. request some "warmup" URL) are a problem because nothing prevents other real, live requests from bombarding the app while it's doing the warmup thing. That's the other very important way in which `IProcessHostPreloadClient` is very different from everything else here.

Within the past year, _many_ of our internal apps still stuck on .NET Framework have been implementing pre-load. I really hope it comes to ASP.NET Core soon.

**Comment by MV10** [comment]
So... two more years. We now have MANY applications that have migrated to ASP.NET Core and we routinely field complaints about startup delays and timeouts on initial request, particularly after patching reboots and similar activities.

`IProcessHostPreloadClient` solves the problem.

I don't know the threshold for the "affected-few" tag but in my company it's minimally hundreds of applications at this point, and thousands more in the pipeline.

(Also the docs are wrong, for some reason it was changed to say it's only for IIS 7 / Windows 7 ... reported [here](https://github.com/dotnet/dotnet-api-docs/issues/12510), our Framework apps do this all over the place on Server 2022. Mentioned just in case anyone considers using that as justification to torpedo this preload approach.)

**Comment by MV10** [comment]
Perhaps this will help someone in enterprise scenarios:

As a mitigation which sort of simulates `IProcessHostPreloadClient`, I wrote a library for our ASP.NET Core apps to return a custom HTTP 5XX code from a simple early pipeline handler during init. An F5 load balancer iRule recognize this 5XX, tags the pool member as "busy" (not "down"; uses the bitfield flag table), redirects to another pool member, and the state flag has a simple TTL which is a low-cost "re-test" interval. Post-init the app makes a single library call to disable the 5XX, and thereafter the pipeline overhead becomes a single bool check. 

Obviously if you don't have an F5 or policy prohibits iRule fiddling, you're still up the creek, but at least we have a work-around (and it's platform-agnostic, even our Java brethren _could_ use it, although my go-to recommendation for them is still "migrate to .NET" 🫡).

**Comment by DamianEdwards** [comment]
I did some digging into both the legacy API and the current ANCM/hosting startup path.

The old API is extremely small:

```csharp
public interface IProcessHostPreloadClient
{
    void Preload(string[] parameters);
}
```

IIS instantiates the configured type and synchronously calls `Preload` before normal request processing. It has no DI, cancellation, or async support, and failures abort preload. It is selected through `serviceAutoStartEnabled`, `serviceAutoStartProvider`, and `<serviceAutoStartProviders>`.

For ASP.NET Core, I don't think we need to introduce an equivalent user-facing interface. `IHostedService.StartAsync` is already a better managed initialization contract: it supports DI, async work, and cancellation. With the default sequential startup behavior in the modern hosting model, user-hosted services are registered before `GenericWebHostService`, so their startup completes before `IISHttpServer` registers its request callbacks.

The actual missing capability is **requestless activation by ANCM**. Today ANCM creates the managed application from `RQ_EXECUTE_REQUEST_HANDLER`; `APPLICATION_MANAGER::GetOrCreateApplicationInfo` and `APPLICATION_INFO::CreateApplication` both require an `IHttpContext`. Consequently some request—including the synthetic request generated by IIS Application Initialization when `preloadEnabled=true`—must enter ANCM before CoreCLR and the ASP.NET Core host start.

A first-class implementation appears technically feasible because IIS exposes `GL_APPLICATION_START` / `OnGlobalApplicationStart(IHttpApplicationStartProvider*)`, which supplies an `IHttpApplication` without requiring a request. The work would roughly be:

1. Register ANCM for `GL_APPLICATION_START` (or otherwise integrate with `serviceAutoStartEnabled`).
2. Refactor ANCM application creation so it can start from `IHttpApplication` without an `IHttpContext`.
3. Start CoreCLR and wait for managed host startup, including `IHostedService.StartAsync`, to complete.
4. Gate incoming requests until startup succeeds or `startupTimeLimit` expires.
5. Correctly propagate startup failures and preserve recycle, shutdown, configuration-change, shadow-copy, and in-/out-of-process behavior.

So my assessment is: **yes, this can reasonably be plumbed through, but it is primarily a medium-sized native ANCM lifecycle feature rather than a new API in the IIS hosting package**. The managed initialization API already exists; the gap is getting ANCM to activate the application without manufacturing an HTTP request and ensuring IIS doesn't admit traffic until that activation completes.

**Comment by MV10** [comment]
That's even better than I'd hoped for.
